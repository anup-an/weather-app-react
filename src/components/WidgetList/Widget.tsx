import React from 'react'

const Widget: React.FC = (): JSX.Element => {
    const cities = ["Turku", "Helsinki", "Oulu"]
    return (
        <div className="grid">
            {cities.map(city => {
                <div className="card">
                    <p>cities</p>
                </div>
            })}
        </div>
    )
}
export default Widget