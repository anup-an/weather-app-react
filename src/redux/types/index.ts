export const SET_SEARCH_KEYWORD = 'SET_SEARCH_KEYWORD'
export const SEARCH_CITIES = 'SEARCH_CITIES'
export const LISTEN_SEARCH_CITIES = 'LISTEN_SEARCH_CITIES'
export const ADD_CITY_WEATHER = 'ADD_CITY_WEATHER'
export const LISTEN_ADD_CITY_WEATHER = 'LISTEN_ADD_CITY_WEATHER'
export const DELETE_CITY_WEATHER = 'DELETE_CITY_WEATHER'

export type CityType = {
    city: string
    country: string
}

export type WeatherType = {
    temperature: number
    city: string
    country: string
    time: string
    condition: string
}
export type SetSearchKeywordAction = {
    type: typeof SET_SEARCH_KEYWORD
    payload: {
        keyword: string
    }
}

export type SearchCitiesAction = {
    type: typeof SEARCH_CITIES
    payload: {
        cities: CityType[]
    }
}

export type AddCityWeatherAction = {
    type: typeof ADD_CITY_WEATHER
    payload: {
        weather: WeatherType
    }
}

export type DeleteCityWeatherAction = {
    type: typeof DELETE_CITY_WEATHER
    payload: {
        weather: WeatherType
    }
}

export type AppAction = SearchCitiesAction | AddCityWeatherAction | DeleteCityWeatherAction | SetSearchKeywordAction

export type AppState = {
    cities: CityType[]
    searchKeyword: string
    weather: WeatherType[]
}

export type CityResponseType = {
    data: {
    id: string,
    wikiDataId: string,
    type: string,
    city: string,
    name: string,
    country: string,
    countryCode: string,
    region: string,
    regionCode: string,
    latitude: number,
    longitude: number,
        population: number
    }[]
}

export type WeatherResponseType = {
    "coord": {
        "lon": number,
        "lat": number
    },
    "weather": [
        {
            "id": number,
            "main": string,
            "description": string,
            "icon": string
        }
    ],
    "base": string,
    "main": {
        "temp": number,
        "feels_like": number,
        "temp_min": number,
        "temp_max": number,
        "pressure": number,
        "humidity": number
    },
    "visibility": number,
    "wind": {
        "speed": number,
        "deg": number
    },
    "rain": {
        "1h": number
    },
    "clouds": {
        "all": number
    },
    "dt": number,
    "sys": {
        "type": number,
        "id": number,
        "country": string,
        "sunrise": number,
        "sunset": number
    },
    "timezone": number,
    "id": number,
    "name": string,
    "cod": number
}



