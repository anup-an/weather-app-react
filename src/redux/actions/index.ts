import { CityType, AppAction, SEARCH_CITIES, WeatherType, ADD_CITY_WEATHER, DELETE_CITY_WEATHER, SET_SEARCH_KEYWORD } from "../types"

export const setSearchKeyword = (keyword: string): AppAction => {
    return {
        type: SET_SEARCH_KEYWORD,
        payload: {
            keyword: keyword 
        }
    }
}

export const searchCities = (cities: CityType[]): AppAction => {
    return {
        type: SEARCH_CITIES,
        payload: {
            cities: cities 
        }
    }
}

export const addCityWeather = (weather: WeatherType): AppAction => {
    console.log(weather)
    return {
        type: ADD_CITY_WEATHER,
        payload: {
            weather: weather 
        }
    }
}

export const deleteCityWeather = (weather: WeatherType): AppAction => {
    return {
        type: DELETE_CITY_WEATHER,
        payload: {
            weather: weather 
        }
    }
}