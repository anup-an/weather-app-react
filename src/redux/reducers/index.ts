import { AppAction, AppState } from '../types';
import * as t from '../types' 

const initialState: AppState = {
    cities: [],
    searchKeyword: '',
    weather: []
}

const reducer = (state = initialState, action: AppAction) => {
    switch (action.type) {
        case t.SET_SEARCH_KEYWORD:
            return { ...state, searchKeyword: action.payload.keyword }
        case t.SEARCH_CITIES:
            return { ...state, cities: [...action.payload.cities] }
        case t.ADD_CITY_WEATHER:
            const filteredList = [...state.weather].filter(weather => weather.city !== action.payload.weather.city)
            return { ...state, weather: [...filteredList, action.payload.weather] }
        case t.DELETE_CITY_WEATHER: 
            const updatedList = [...state.weather].filter(weather => weather.city !== action.payload.weather.city)
            return { ...state, weather: [...updatedList] }
        default:
            return state
    }
    
}

export default reducer

