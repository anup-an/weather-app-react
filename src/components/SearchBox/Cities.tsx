import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { AppState, LISTEN_ADD_CITY_WEATHER } from '../../redux/types'

const Cities: React.FC = (): JSX.Element => {
    const { cities } = useSelector((state: AppState) => state)
    const dispatch = useDispatch()
    const addWeatherWidget = (city: string, country: string) => {
        dispatch({type: LISTEN_ADD_CITY_WEATHER, name: city, country: country})
    }
    return (
        <div className='list_container'>
            <ul className="cities_ul">
                {cities.map(city => (
                    <li key={city.city}>
                        <p>{city.city},{' '}{city.country}</p>
                        <button onClick={() => addWeatherWidget(city.city, city.country)}>Add +</button>
                    </li>
                ))}
            </ul>
        </div>
    )
}

export default Cities