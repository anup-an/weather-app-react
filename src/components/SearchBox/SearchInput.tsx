import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { useDebounce } from '../../hooks';
import { searchCities, setSearchKeyword } from '../../redux/actions';
import { AppState, LISTEN_SEARCH_CITIES } from '../../redux/types';
import '../../styles/App.css'

const SearchInput: React.FC = (): JSX.Element => {
    const dispatch = useDispatch()
    const {searchKeyword} = useSelector((state: AppState) => state)
    const debouncedKeyword = useDebounce(searchKeyword, 2000);

    const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault()
        dispatch({type: LISTEN_SEARCH_CITIES, keyword: searchKeyword})
    }

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        dispatch(setSearchKeyword(event.target.value))
    }

    useEffect(
        () => {
            if (debouncedKeyword && debouncedKeyword !== '') {
                console.log(debouncedKeyword)
                dispatch({ type: LISTEN_SEARCH_CITIES, keyword: debouncedKeyword })               
            } else {
                dispatch(searchCities([]))
            }
        }, [debouncedKeyword]
    );
    
    return (
        <div className='search_container'>
        <form onSubmit={handleSubmit} className='search-form'>        
            <label htmlFor="keywords">
                <input
                    id="keywords"
                    name="keywords"
                    type="text"
                    placeholder="Search for a city"
                        className="input"
                        onChange={handleInputChange}
                    size={100}
                />
                        </label>
                <button className="search-button">

                <svg
                    xmlns="http://www.w3.org/2000/svg"
                    className='icon'
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                >
                <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                    d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
                />
                        </svg> 
                </button>

            
            </form>
            </div>
    )
}

export default SearchInput