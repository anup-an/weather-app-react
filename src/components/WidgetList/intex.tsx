import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { deleteCityWeather } from '../../redux/actions'
import { AppState, WeatherType } from '../../redux/types'

const WidgetList: React.FC = (): JSX.Element => {
    const { weather } = useSelector((state: AppState) => state)
    const dispatch = useDispatch()
    const deleteWidget = (cityWeather: WeatherType) => {
        dispatch(deleteCityWeather(cityWeather))
    }
    return (
        <ul className="grid">
            {weather.map(cityWeather => (
                <li className="card">
                    <button onClick={() => deleteWidget(cityWeather)}>
                        <p>x</p>
                    </button>
                    <div>
                        <div>
                            <p>{cityWeather.temperature}</p>
                            <span id="celcius_symbol">&#8451;</span>
                        </div>                       
                        <div>
                            <p>{cityWeather.city}</p>
                            <p>{cityWeather.country}</p>
                            <p>{new Date().getHours()}:{new Date().getMinutes()}</p>
                        </div>
                    </div>
                    <p>{cityWeather.condition}</p>
                    
                </li>
            ))}
        </ul>
    )
}

export default WidgetList