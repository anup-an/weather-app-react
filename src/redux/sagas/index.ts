import { call, put, takeLatest, takeEvery } from "@redux-saga/core/effects"
import { all, select } from 'redux-saga/effects'
import axios, { AxiosResponse } from "axios"
import { addCityWeather, searchCities } from "../actions"
import { ADD_CITY_WEATHER, AppAction, AppState, CityResponseType, CityType, DELETE_CITY_WEATHER, LISTEN_ADD_CITY_WEATHER, LISTEN_SEARCH_CITIES, WeatherResponseType, WeatherType} from "../types"
import { Action } from "redux"

export type CityInputType = {
    keyword: string
}

export type WeatherInputType = {
    name: string
    country: string
}
export type AppActionType = Action & AppAction & CityInputType & WeatherInputType

export function* searchCitiesSaga(search: CityInputType) {
    const { keyword } = search
    try {
        if (keyword !== '') {
            const response: AxiosResponse<CityResponseType> = yield call(axios.get, `http://geodb-free-service.wirefreethought.com/v1/geo/cities?namePrefix=${keyword}`)
            const cities: CityType[] = response.data.data.map(foundCity => ({ city: foundCity.name, country: foundCity.country })) 
            yield put(searchCities(cities))
        } else {
            yield put(searchCities([]))
        }
    } catch (error) {
        console.log(error)
    }
}

export function* addWeatherSaga(city: WeatherInputType) {
    const {name, country} = city
    try {    
        const response: AxiosResponse<WeatherResponseType> = yield call(axios.get, `http://api.openweathermap.org/data/2.5/weather?q=${name}&units=metric&appid=1e9e35761dc83e1c61236cc9333bd5fe`)
        const weather: WeatherType = {
            city: response.data.name,
            country: country,
            temperature: Math.round(response.data.main.temp),
            time: new Date().getHours() + ':' + new Date().getMinutes(),
            condition: response.data.weather[0].main
        }
        yield put(addCityWeather(weather))
    } catch (error) {
        console.log(error)
    }
}

export function* saveWeatherWithSaga() {
  try {
    const state: AppState = yield select()
    yield localStorage.setItem('weather', JSON.stringify(state.weather))
  } catch (error) {
    console.log(error)
  }
}

export function* searchCitiesActionWatcher() {
    yield takeLatest<AppActionType>(LISTEN_SEARCH_CITIES, searchCitiesSaga)
}

export function* addWeatherActionWatcher() {
    yield takeEvery<AppActionType>(LISTEN_ADD_CITY_WEATHER, addWeatherSaga)
}

export function* saveWeatherActionWatcher() {
    yield takeEvery<AppActionType>([ADD_CITY_WEATHER, DELETE_CITY_WEATHER], saveWeatherWithSaga)
}

export default function* rootSaga() {
    yield all([
        takeLatest<AppActionType>(LISTEN_SEARCH_CITIES, searchCitiesSaga),
        takeEvery<AppActionType>(LISTEN_ADD_CITY_WEATHER, addWeatherSaga),
        takeEvery<AppActionType>([ADD_CITY_WEATHER, DELETE_CITY_WEATHER], saveWeatherWithSaga)
    ])
}