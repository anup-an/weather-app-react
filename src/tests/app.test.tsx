import { mount } from 'enzyme'
import { Provider } from 'react-redux';
import Cities from '../components/SearchBox/Cities';
import makeStore from '../redux/store';
import SearchInput from '../components/SearchBox/SearchInput';
import WidgetList from '../components/WidgetList';
import { Store } from 'redux';
import { AppState, AppAction } from '../redux/types';

describe('Components', () => {
    let store: Store<AppState, AppAction> & {dispatch: unknown};
    beforeEach(() => {
        store = makeStore({
            cities: [{
                city: 'Kathmandu',
                country: 'Nepal',
            }],
            weather: [{
                temperature: 5,
                city: 'Kathmandu',
                country: 'India',
                time: '20:30',
                condition: 'Cloudy'
            }],
            searchKeyword: 'Kathmandu'
        })
    });
    it('Snapshot testing for <Cities> component', () => { 
    
        
        const wrapper = mount(
            <Provider store={store}>
                <Cities />
            </Provider>
        );
      
        expect(wrapper).toMatchSnapshot();
    })
    
    it('Snapshot testing for <SearchInput/> component', () => { 
            
        const wrapper = mount(
            <Provider store={store}>
                <SearchInput/>
            </Provider>
        );
        
    expect(wrapper).toMatchSnapshot();
    })

    it('Snapshot testing for <WidgetList/> component', () => { 
        const wrapper = mount(
            <Provider store={store}>
                <WidgetList/>
            </Provider>
        );
        
    expect(wrapper).toMatchSnapshot();
    })

})