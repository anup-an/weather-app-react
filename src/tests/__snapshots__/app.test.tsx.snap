// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Snapshot testing for <Cities> component 1`] = `ReactWrapper {}`;

exports[`Components Snapshot testing for <SearchInput/> component 1`] = `ReactWrapper {}`;

exports[`Components Snapshot testing for <WidgetList/> component 1`] = `ReactWrapper {}`;
